source/loader/lib_loader.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+#ifndef LIB_LOADER_HPP
+#define LIB_LOADER_HPP
+
+#include <array>
+
+#include "logger/ur_logger.hpp"
+#include "ur_api.h"
+#include "ur_ddi.h"
+#include "ur_object.hpp"
+
+namespace loader {
+
+using LibHandle = std::unique_ptr<void, void (*)(HMODULE)>;
+
+struct platform_t {
+    platform_t(LibHandle handle) : handle(std::move(handle)) {}
+
+    HMODULE getFunctionPtr(const char *func_name);
+
+    LibHandle handle;
+    ur_result_t initStatus = UR_RESULT_SUCCESS;
+    dditable_t dditable = {};
+};
+
+using platform_vector_t = std::vector<platform_t>;
+
+class LibLoader {
+  public:
+    void loadLibraries(platform_vector_t &platforms) {
+        discoverEnabledPlatforms();
+        for (const auto &name : discoveredPlatforms) {
+            auto handle = loadAdapterLibrary(name.c_str());
+            if (handle) {
+                platforms.emplace_back(std::move(handle));
+            }
+        }
+    }
+
+  private:
+    static constexpr std::array<char *, 0> knownPlatformNames{};
+    std::vector<std::string> discoveredPlatforms;
+
+    static LibHandle loadAdapterLibrary(const char *name);
```

Comment:
It would be cleaner to create a standalone abstraction (with its own class) for this in common. And then the loader can just instantiate that class to access the functions.

Reply:
Split the code into three parts: loading libraries, finding adapters, and platform_t structure. Moved first two abstractions to common.